core/che-core-api-core/src/main/java/org/eclipse/che/api/core/util/PagingUtil.java

```diff
@@ -0,0 +1,106 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2016 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.core.util;
+
+import org.eclipse.che.api.core.Page;
+import org.eclipse.che.commons.lang.Pair;
+
+import javax.ws.rs.core.UriBuilder;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.lang.String.format;
+import static java.util.Collections.emptyMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+
+/**
+ * Provides useful methods for working with {@link Page} instances
+ * and pageable uris management.
+ *
+ * @author Yevhenii Voevodin
+ */
+public final class PagingUtil {
+
+    private static final String  LINK_HEADER_SEPARATOR = ", ";
+    /**
+     * Helps to retrieve href along with rel from link header value part.
+     * Value format is {@literal <http://host:port/path?query=value>; rel="next"},
+     * so the first group is href while the second group is rel.
+     */
+    private static final Pattern LINK_HEADER_REGEX     = Pattern.compile("<(.+)>;.*rel=\"(.+)\".*");
+
+    /**
+     * Generates link header value from the page object and base uri.
+     * <a href="https://tools.ietf.org/html/rfc5988">The Link header spec</a>
+     *
+     * @param page
+     *         the page used to generate link
+     * @param uri
+     *         the uri which is used for adding {@code skipCount} & {@code maxItems} query parameters
+     * @return 'Link' header value
+     * @throws NullPointerException
+     *         when either {@code page} or {@code uri} is null
+     */
+    public static String createLinkHeader(Page<?> page, URI uri) {
+        requireNonNull(page, "Required non-null page");
+        requireNonNull(uri, "Required non-null uri");
+        final ArrayList<Pair<String, Page.PageRef>> pageRefs = new ArrayList<>(4);
```

Comment:
As it is short living collection and in case when there is next and previous pages it is unneeded to create collection that will grow. So please increase the size of this collection. In case when there is no next/prev pages this collection will be ripped fast enough.